redaxo/src/core/lib/login/password_policy.php

```diff
@@ -0,0 +1,104 @@
+<?php
+
+/**
+ * @author gharlan
+ *
+ * @package redaxo\core
+ */
+class rex_password_policy
+{
+    use rex_factory_trait;
+
+    private $options;
+
+    protected function __construct(array $options)
+    {
+        $this->options = $options;
+    }
+
+    /**
+     * @return static
+     */
+    public static function factory(array $options)
+    {
+        $class = static::getFactoryClass();
+
+        return new $class($options);
+    }
+
+    public function check($password)
```

Comment:
ich vermute einige passwort policies brauchen auch noch den aktuellen user/user-id (da man ggf. pro user ein passwort nicht "wiedervenden darf" o.ä.)

Reply:
diese methode (da API) bitte mit phpdoc

Reply:
Guter Punkt. Wobei die Klasse nicht nur fürs Backend gedacht ist, sondern auch für andere Stellen im Frontend etc, wo man Passwörter checken will.
Aber ich denke ich füge trotzdem einfach einen $id-Parameter hinzu. Und im backend wird halt die User-ID des Backend-Users übergeben.

Reply:
werden die $options fürs frontend dann auch in der config.yml konfiguriert oder kommen die dann woanders her?

Reply:
fürs festlegen der policy fürs frontend wird dann vermutlich auch nicht die rex-factory benutzt, oder?

Reply:
Die Regeln werden an factory übergeben, somit kann man sich im Frontend aussuchen, ob man auch die `password_policy`-Property abfragt und übergibt, oder ob man was anderes übergibt.
Auch bzgl factory kann man es sich aussuchen, ob man factory nutzt, oder direkt eine eigene Kind-Klasse.
Allerdings kann man aktuell nicht explizit diese Basisklasse nehmen. Das ist vielleicht nicht so gut. Werde daher den Constructor public machen, sodass das auch möglich ist.

Reply:
warum ich so komisch gefragt hab: es sollte sichergestellt sein dass man nicht im backend regeln konfiguriert und diese dann "ausversehen" im frontend greifen oder umgekehrt (falls es wirklich über die factory läuft, da es ja nur 1 impl geben kann pro factory)

Reply:
ggf. wäre es sinnvoll 3 klassen zu haben

```php
abstract class password_policy {}  // ohne factory
class backend_password_policy extends password_policy{} // backend factory
class frontend_password_policy extends password_policy{} // frontend factory
```

Reply:
Frage ist auch wichtig und berechtigt. Ich denke aber, dass es nun passt.
Habe bewusst nicht innerhalb der Klasse die Property aus der config.yml ausgelesen.

Reply:
Finde ich nicht notwendig mit den Extra-Klassen.

Reply:
wenn man diese klassen nicht hat muss man vorm registrieren der factory immer prüfen wo man gerade ist, dass man via addon nicht "ausversehen" die backend policy ändert (oder ausversehen die frontend policy).

ausserdem setzt eine factory vorraus dass innerhalb eines requests immer nur frontend oder backend code läuft, niemals beides.

Reply:
Ich kann nicht richtig folgen.

Ich sehe es so: Das Backend nutzt an den Stellen für die Backend-User die factory.
Wer sich da rein hängen will, setzt seine eigene Klasse per setFactoryClass.

Andere Stellen, egal ob Backend oder Frontend, wo man selbst, aus welchen Gründen auch immer, ein Passwort gemäß einer Policy checken will, sind frei in der Wahl, ob sie sich der Factory für die Backend-user anschließen, oder eine eigene Klasse direkt, oder die Basisklasse direkt nutzen wollen.
Ich sehe es eher so, dass ycom zum beispiel eine eigene Basisklasse anbieten kann. Aber nicht der Core allgemein fürs Frontend. Es kann auch im Frontend nochmal unterschiedliche Zwecke geben.

Reply:
angenommen ein addon will eine policy für frontend-user definieren.
- wenn dafür die rex-factory verwendet würde würde automatisch auch die backend policy geändert (was vermutlich nie gewollt ist)
- wenn im backend eine userverwaltung für frontend user gäbe und die policy via rex-factory registriert werden würde müsste man einiges an gehirn jogging aufwänden damit man dann trotz backend die richtige frontend policy beim ändern eines frontend users zum tragen kommt

Reply:
Kompromissvorschlag:
`rex_password_policy` als Basisklasse, ohne Factory, aber auch nicht abstrakt.
Und `rex_backend_password_policy extends rex_password_policy ` mit Factory.